public/app/features/explore/RichHistory/RichHistoryCard.tsx

```diff
@@ -147,28 +151,37 @@ export function RichHistoryCard(props: Props) {
   } = props;
   const [activeUpdateComment, setActiveUpdateComment] = useState(false);
   const [comment, setComment] = useState<string | undefined>(query.comment);
+  const [queryDsInstance, setQueryDsInstance] = useState<DataSourceApi | undefined>(undefined);
+
+  useEffect(() => {
+    getQueryDsInstance();
+  });
 
-  const toggleActiveUpdateComment = () => setActiveUpdateComment(!activeUpdateComment);
   const theme = useTheme();
   const styles = getStyles(theme, isRemoved);
 
+  const getQueryDsInstance = async () => {
+    const ds = await getDataSourceSrv().get(query.datasourceName);
+    setQueryDsInstance(ds);
+  };
+
   const onRunQuery = async () => {
-    const dataQueries = query.queries.map((q, i) => createDataQuery(query, q, i));
+    const queriesToRun = query.queries.map((q, i) => createDataQuery(query, q, i));
     if (query.datasourceName !== datasourceInstance?.name) {
       await changeDatasource(exploreId, query.datasourceName);
-      setQueries(exploreId, dataQueries);
+      setQueries(exploreId, queriesToRun);
     } else {
-      setQueries(exploreId, dataQueries);
+      setQueries(exploreId, queriesToRun);
     }
   };
 
   const onCopyQuery = () => {
-    const queries = query.queries.join('\n\n');
-    copyStringToClipboard(queries);
+    const queriesToCopy = query.queries.map(q => createQueryText(q, queryDsInstance)).join('\n');
+    copyStringToClipboard(queriesToCopy);
     appEvents.emit(AppEvents.alertSuccess, ['Query copied to clipboard']);
   };
 
-  const onCreateLink = () => {
+  const onCreateLink = async () => {
```

Comment:
Does this need to be async? Or is this missing `await` somewhere?

Reply:
Yeah, async is redundant, I accidentally left it there. Good catch.